Stream index files for fetch metadata calls and serve reads without series block metadata in memory

robskillington opened the pull request:
Notes:
- [x] Need to move out stitching series data and series index metadata together from series itself
- [x] Make the V2 endpoint work even with CacheAll and CacheAllMetadata series cache policies so that old clusters can be upgraded
- [x] Need to move the management of the index readers from shard to some other manager that can manage cache and reuse of the readers
- [x] Validate index data file before streaming metadata back to peers

Todo with a follow up change:
- Cache blocks as they are fetched so that subsequent requests read them directly from memory
- Once added, add metrics to the series reader to check how often cache misses occur for reads

richardartoul commented on storage/namespace_readers.go:
```diff
@@ -0,0 +1,313 @@
+		err    error
+	)
+	if len(m.closedReaders) > 0 {
+		idx := len(m.closedReaders) - 1
```
super nit: might be nice to wrap these lines into a pop() method and the append(m.closedReader, ...) in put into a push() method. That way its much easier to realize that you're treating m.closedReaders as a stack

robskillington replied:
Done, ta.

richardartoul commented on storage/namespace_readers.go:
```diff
@@ -0,0 +1,313 @@
+		return nil, err
+	}
+
+	// Fast fwd through if in the middle of a volume
```
This is extremely unlikely to happen because a given peer is never making parallel calls to a server within a single shard/block, right?

robskillington replied:
Correct yeah, this is for the special case where client resumes a lot later.

richardartoul commented on storage/namespace_readers.go:
```diff
@@ -0,0 +1,313 @@
+	defer m.Unlock()
+
+	if !status.Open {
+		m.closedReaders = append(m.closedReaders, cachedReader{
```
same comment as before about push/pop

robskillington replied:
Sure thing, will add the two methods.

richardartoul commented on storage/namespace_readers.go:
```diff
@@ -0,0 +1,313 @@
+	}
+	// Expire any closed readers, alloc a new slice to avoid spikes
+	// of use creating slices that are never released
+	if expired := expiredClosedReaders; expired > 0 {
```
Can this just be

```golang
if expiredClosedReaders > 0 {
...
}
```

robskillington replied:
Yes but I use the `expired` symbol below to keep under the 80 char limit for a line or two.. ha.

richardartoul commented on storage/namespace_readers.go:
```diff
@@ -0,0 +1,313 @@
+		}
+	}
+	// Expire any closed readers, alloc a new slice to avoid spikes
+	// of use creating slices that are never released
```
Not sure what you're worried about here in terms of allocation, how would you create slices that are never released?

richardartoul replied:
Also I'm not sure why you need to expire the closed readers. The get/put logic should handle re-using them for you automatically right? Are you just trying to prevent the slice size from growing too large after a burst and never shrinking?

robskillington replied:
The latter yes, preventing the slice from growing large after a burst and never releasing the memory back.

richardartoul commented on storage/namespace_readers.go:
```diff
@@ -0,0 +1,313 @@
+
+	// For open readers calculate and expire from map directly
+	for key, elem := range m.openReaders {
+		elem.ticksSinceUsed++
```
the ++ is confusing because it makes it seem like you inadvertently introduced a bug by trying to mutate a copy. I can't think of a way to make the code more clear, so maybe a comment here saying "mutates the copy, but not the element in the map itself" and then on line 311 where you reassign "reassign the mutated copy"

robskillington replied:
Sure thing.

richardartoul commented on storage/namespace_readers.go:
```diff
@@ -0,0 +1,313 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
```
Gettin' dangerously close to implementing sessions here ;)

A file-level or struct-level comment explaining what a databaseNamespaceReaderManager is and why its important would be great. Also might be nicer to rename to it databaseNamespaceReaderCache because primary purpose seems to be caching not lifecycle management (it just happens to do that as a side-effect of caching)

> databaseNamespaceReaderManager maintains a pool of closed readers which can be re-used (to prevent additional allocations), as well as a cache of recently used open readers based on their position. The cache of recently used open readers is useful during peer bootstrapping because a pageToken (which contains an offset into the reader for both the data and metadata portions of the fileset) is used to communicate the clients current position to the server. In the general case, the client will miss on its first request for a given shard/block start, and then experience a cache hit on every subsequent request because the current client implementation does not perform any parallel requests within a given shard.
> 
> The closedReaders pool is modeled as a stack (implemented via slice operations) and the open readers cache is implemented as a map where the key is of type cachedOpenReaderKey.
> 
> the databaseNamespaceReaderManager also implements a tick() method which should be called regularly in order to shrunk the closedReaders stack after bursts of usage, as well as to expire cached open readers which have not been used for a configurable number of ticks.

robskillington replied:
Haha yes, this was almost a sessions implementation.  Thankfully if a cache miss occurs the caller doesn't have to start all the way from the start again however.

Thanks for the writeup, very useful.  I'll add it to the file.  I mainly skip these in the interest of time, but we should definitely be adding them.

richardartoul commented on storage/series/policy.go:
```diff
@@ -0,0 +1,106 @@
+	// TODO: Remove this once recently read is production grade.
+	CacheAllMetadata
+	// CacheRecentlyRead specifies that series that are recently read
+	// must be cached, configurable by the namespace block expiry after
```
Could you elaborate "configurable by the namespace block expiry after not accessed period"? I'm not 100% sure what this means

robskillington replied:
It's a field that's configurable per namespace, is it not totally clear here?

richardartoul commented on storage/series/policy.go:
```diff
@@ -0,0 +1,106 @@
+
+const (
+	// DefaultCachePolicy is the default cache policy.
+	DefaultCachePolicy = CacheRecentlyRead
```
Could this be moved down into the same const block where the actual policies are defined (preferably at the end?)

robskillington replied:
Sure thing.

richardartoul commented on storage/series/policy.go:
```diff
@@ -0,0 +1,106 @@
+)
+
+// CachePolicy is the series cache policy.
+type CachePolicy uint
```
Whats the benefit of adding this extra type def instead of just using iota directly

robskillington replied:
I need it to be a type so I can add a `String()` method to it.

richardartoul commented on storage/series/policy.go:
```diff
@@ -0,0 +1,106 @@
+)
+
+// ValidCachePolicies returns the valid series cache policies.
+func ValidCachePolicies() []CachePolicy {
```
super nit: In the session file we just have a global slice. Would be nice to be consistent in terms of having a function that returns it, or a global slice

robskillington replied:
I use a function to make it impossible for caller to mutate it.  I can update sessions usage, although its only internal to the package there..

richardartoul commented on storage/series/reader.go:
```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
```
2017

robskillington replied:
Haha, ta.

richardartoul commented on storage/series/reader.go:
```diff
@@ -0,0 +1,250 @@
+	seriesBlocks block.DatabaseSeriesBlocks,
+	seriesBuffer databaseBuffer,
+) ([][]xio.SegmentReader, error) {
+	// TODO(r): pool these results arrays
```
Is this TODO a production blocker?

robskillington replied:
Nope, we live with this today in production (ha).  Once we start doing 1MM read / sec we probably will want to do things like this.

richardartoul commented on storage/series/reader.go:
```diff
@@ -0,0 +1,250 @@
+	retriever   QueryableBlockRetriever
+}
+
+// NewReaderForRetriever returns a reader for a series
```
I'm a little confused why we need this new reader. Aren't we currently serving reads from these block retrievers already? Why do we need something new to support peer bootstrapping?

robskillington replied:
So this was because we don't want to create a `Series` when we want to service a read for a series that isn't in memory.  It's moving the stitching of results together from the `Series` to an independent party that can be used directly in the shard (without needing to create a `Series` when one doesn't exist in memory).

richardartoul replied:
Ok cool, possibly worth a comment, but maybe not. Kind of requires a wholistic understanding of whats going on

richardartoul commented on storage/series/reader.go:
```diff
@@ -0,0 +1,250 @@
+	return r.fetchBlocksWithBlocksMapAndBuffer(ctx, starts, nil, nil)
+}
+
+func (r Reader) fetchBlocksWithBlocksMapAndBuffer(
```
Similar to my comment above, a little confused as to what this function is for and why these access pattern is different than what we do to serve reads

robskillington replied:
As mentioned above, this was because we don't want to create a Series when we want to service a read for a series that isn't in memory. It's moving the stitching of results together from the Series to an independent party that can be used directly in the shard (without needing to create a Series when one doesn't exist in memory).

richardartoul commented:
Also if you could update the P.R summary before landing that would be really helpful if we ever need to look back at this P.R (plus it'll be easier to think about what this P.R does at a high level and if there is anything we missed)

robskillington commented:
Re: general questions:
> You switched the Write() method to Update(). I had that in the back of my mind while I reviewed, but I must have missed where you made that change. What happened there?

It seemed more pragmatic to use Update since its updating a digest and returning a new result vs Write which implies its writing more data to the digest, which is not true because the digest is not mutable anymore and returns a new digest when wanting to add more contents of the digest.

> Why did you delete the last_read_after_peer_boostrap integration test?

I deleted it because the last read field is not going to be propagated any longer.  We actually need to remove the whole concept from being accessible from RPCs any longer, we don't persist this and won't in the future so it's no longer something we want to test / can test.

> In my P.R I introduced the concept of running all the integration tests for two different mem policies (see the first two files here: m3db/m3db/pull/411/files)
> Seems to me like we should run every integration test with each of the series caching policies

True, once I rebase I'll try and do this.  It is a little more complicated than this because some tests won't be able to run with both (i.e. for the integration tests that are testing v1 fetch blocks metadata endpoints the newer caching policies aren't compatible, so will need to find a way to run that/ not run that dependent on the caching policy).

robskillington commented on client/session.go:
```diff
@@ -78,16 +77,22 @@ const (
+	FetchBlocksMetadataEndpointV1
 	// FetchBlocksMetadataEndpointV2 represents v2 of the fetch blocks metadata endpoint
-	FetchBlocksMetadataEndpointV2
+	FetchBlocksMetadataEndpointV2 = FetchBlocksMetadataEndpointDefault
```
I can't unfortunately since the default needs to be a zero value (as there's a lot of place where it's not set by default and we want the default to actual be the new endpoint).

robskillington commented on client/session.go:
```diff
@@ -1978,24 +1980,33 @@ func (s *session) emitDuplicateMetadataLog(received *receivedBlocks, metadata bl
 
 func (s *session) emitDuplicateMetadataLogV2(received *receivedBlocks, metadata blocksMetadata) {
-	fields := make([]xlog.Field, len(received.results)+1)
+	fields := make([]xlog.Field, 0, len(received.results)+1)
```
Yup, I'll merge in your fixes.

robskillington commented on client/session.go:
```diff
@@ -3183,3 +3189,32 @@ func IsValidFetchBlocksMetadataEndpoint(endpointVersion FetchBlocksMetadataEndpo
+			return nil
+		}
+	}
+	return fmt.Errorf("invalid series CachePolicy '%s' valid types are: %v",
```
Good catch, ta.

robskillington commented on digest/digest.go:
```diff
@@ -21,31 +21,33 @@
-	d := NewDigest()
-	if head := segment.Head; head != nil {
-		d.Write(head.Get())
+	d := stackadler32.NewDigest()
```
So `NewDigest()` is now just a struct rather than a heap allocated instance, which is a significant savings.

richardartoul replied:
Not sure I'm following, I'm just saying line 40 should be equivalent to 

```golang
d := NewDigest()
```

robskillington commented on digest/digest.go:
```diff
@@ -21,31 +21,33 @@
+
+// Checksum returns the checksum for a buffer.
+func Checksum(buf []byte) uint32 {
+	return adler32.Checksum(buf)
```
Not really, but mainly so that only `digest` package needs to know about which implementation we're using - that way if we ever update to something other than adler32 we only need to change all references to adler32 in this one package (encapsulation)

robskillington commented on digest/fd_digest.go:
```diff
@@ -48,7 +49,7 @@ type fdWithDigest struct {
 func newFdWithDigest() FdWithDigest {
 	return &fdWithDigest{
-		digest: NewDigest(),
+		digest: adler32.New(),
```
Since these aren't highly concurrent and we cache the use of them, we don't have to use the stack based adler32 approach - we can use the stdlib which is a few percent faster than the stack based adler32 approach (although stack based is zero-alloc, it'll only be fast in the highly concurrent use case).

richardartoul replied:
Cool, this probably warrants a comment somewhere in this package

robskillington replied:
Yeah fair, I'll add some comments.

robskillington commented on digest/reader.go:
```diff
@@ -180,7 +181,7 @@ type readerWithDigest struct {
 	return &readerWithDigest{
 		reader: reader,
-		digest: NewDigest(),
+		digest: adler32.New(),
```
Same here.  Since these aren't highly concurrent and we cache the use of them (in the readers, seekers, etc), we don't have to use the stack based adler32 approach - we can use the stdlib which is a few percent faster than the stack based adler32 approach (although stack based is zero-alloc, it'll only be fast in the highly concurrent use case).

robskillington commented on generated/mocks/writer/writer_main.go:
```diff
@@ -0,0 +1,95 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
```
This is for fixing the silly mockgen output that generates self-package referencing code for some mocks, this will strip it.  Now we don't have to do this by hand everytime we generate mocks.

robskillington commented on generated/proto/pagetoken/pagetoken.proto:
```diff
@@ -2,5 +2,14 @@ syntax = "proto3";
+	}
+	message FlushedSeriesPhase {
+		int64 currBlockStartUnixNanos = 1;
+		int64 currBlockIndexRead = 2;
```
It's how many entries into the current index pointed at by UnixNanos should be resumed from.  i.e. how far through an file you were half way through to resume from.  

I can make it simpler, however not sure what is a better name.

richardartoul replied:
maybe just remove "read" from the end of the name, then its clear you can just use it as an offset

robskillington replied:
Going with `currBlockEntryIdx`

robskillington commented on integration/integration.go:
```diff
@@ -180,6 +190,12 @@ func newDefaultBootstrappableTestSetups(
+		// Force correct series cache policy if using V1 version
+		// TODO: Remove once v1 endpoint is gone
+		if anySetupUsingFetchBlocksMetadataEndpointV1 {
+			setup.storageOpts = setup.storageOpts.SetSeriesCachePolicy(series.CacheAll)
```
Sure, although we should just remove this asap once v1 is gone.

robskillington commented on integration/setup.go:
```diff
@@ -366,7 +385,7 @@ func (ts *testSetup) waitUntilServerIsDown() error {
 	log := ts.storageOpts.InstrumentOptions().Logger()
 	fields := []xlog.Field{
-		xlog.NewField("nativepooling", ts.nativePooling),
+		xlog.NewField("cachepolicy", ts.storageOpts.SeriesCachePolicy().String()),
```
Considering we do not run native pooling in any tests/anywhere by default anymore I don't see the reason to - it also takes up significant space in the logs since this is attached to every log line.

richardartoul replied:
fair enough

robskillington commented on network/server/tchannelthrift/convert/convert.go:
```diff
@@ -99,33 +100,45 @@ func ToTimeType(unit xtime.Unit) (rpc.TimeType, error) {
 
 	s := &rpc.Segments{}
 
 	if len(readers) == 1 {
```
Correct yes, it let's the caller know that it is fully merged.

robskillington commented on network/server/tchannelthrift/node/service.go:
```diff
@@ -352,22 +349,21 @@ func (s *service) FetchBlocksRaw(tctx thrift.Context, req *rpc.FetchBlocksRawReq
-				block.Checksum = &cki
-			}
-			if err := fetchedBlock.Err(); err != nil {
+			block.Start = xtime.ToNanoseconds(fetchedBlock.Start)
```
Then I'd have to cast to `xtime.UnixNano` though no?

robskillington replied:
Right yeah I thought this was a field with type `xtime.UnixNano` not int64.  Updating.

robskillington commented on network/server/tchannelthrift/node/service.go:
```diff
@@ -352,22 +349,21 @@ func (s *service) FetchBlocksRaw(tctx thrift.Context, req *rpc.FetchBlocksRawReq
-			}
-			if err := fetchedBlock.Err(); err != nil {
+			block.Start = xtime.ToNanoseconds(fetchedBlock.Start)
+			if err := fetchedBlock.Err; err != nil {
```
Yeah so basically this guards that the transmission of the block does not fault, which before we added the ref counting everywhere was a source of risk/concern.

robskillington commented on persist/fs/read.go:
```diff
@@ -351,19 +379,29 @@ func (r *reader) entryID(id []byte) ts.ID {
+	return multiErr.FinalError()
+}
+
+// NB(r): ValidateMetadata can be called as soon as open since the metadata
```
Sure thing, thanks.

robskillington commented on persist/fs/read.go:
```diff
@@ -382,28 +420,55 @@ func (r *reader) EntriesRead() int {
+	decoder := r.decoder
+	digestBuf := r.digestBuf
+	bytesPool := r.bytesPool
+	indexEntriesByOffsetAsc := r.indexEntriesByOffsetAsc
```
I re-zero and set the slice to zero length above this line, hence I reset it back on the struct as we want to reuse the backing array (avoid realloc'ing this slice if we reuse this reader).

richardartoul replied:
Oh I see now

robskillington commented on services/m3dbnode/config/bootstrap.go:
```diff
@@ -39,6 +38,10 @@ import (
+
+	// defaultFetchBlocksMetadataEndpointVersion is the default client fetch blocks
+	// metadata endpoint version to use
+	defaultFetchBlocksMetadataEndpointVersion = client.FetchBlocksMetadataEndpointV2
```
Good call, will do, ta.

robskillington commented on services/m3dbnode/config/bootstrap.go:
```diff
@@ -48,7 +51,27 @@ type BootstrapConfiguration struct {
+	return int(math.Ceil(float64(runtime.NumCPU()) * np))
+}
+
+// TODO: Remove once v1 endpoint no longer required.
```
Maybe we just search everywhere for v1?

robskillington commented on storage/block/block.go:
```diff
@@ -195,6 +197,21 @@ func (b *dbBlock) IsRetrieved() bool {
+	return wasRetrieved
+}
+
+func (b *dbBlock) IsCachedBlock() bool {
```
The comment is on the interface type, I try to avoid having the same text in two places:

richardartoul replied:
for sure, didn't see the interface comment

robskillington commented on storage/bootstrap/bootstrapper/fs/source.go:
```diff
@@ -311,11 +318,13 @@ func (s *fileSystemSource) loadShardReadersDataIntoShardResult(
 				err         error
 			)
-			if retriever == nil {
+			switch seriesCachePolicy {
```
Yeah, I can return an error here.  Basically you should never be reading files if you are using RecentlyRead, we return early and this code is never exercised if so.

robskillington commented on storage/bootstrap/bootstrapper/fs/source.go:
```diff
@@ -426,6 +444,36 @@ func (s *fileSystemSource) Read(
 		}
 	}
 
+	switch s.opts.ResultOptions().SeriesCachePolicy() {
```
Thing is we're going to have other caching policies like RecentlyRead that should be treated like RecentlyRead.  (e.g. LRU based caching policies)

richardartoul replied:
makes sense

robskillington commented on storage/bootstrap/bootstrapper/fs/source.go:
```diff
@@ -426,6 +444,36 @@ func (s *fileSystemSource) Read(
+				md.ID().String())
+		}
+	default:
+		// Unless we're caching all series in memory, we return just the availability
```
That is better, thanks - will update.

robskillington commented on storage/bootstrap/bootstrapper/fs/source_test.go:
```diff
@@ -160,14 +163,14 @@ func validateTimeRanges(t *testing.T, tr xtime.Ranges, expected []xtime.Range) {
 }
 
 func TestAvailableEmptyRangeError(t *testing.T) {
```
Yeah, I think I do need to add some tests for the other caching policies.

robskillington commented on storage/bootstrap/bootstrapper/peers/source.go:
```diff
@@ -250,11 +252,16 @@ func (s *peersSource) incrementalFlush(
-		blockSize      = ropts.BlockSize()
-		shardRetriever = shardRetrieverMgr.ShardRetriever(shard)
-		tmpCtx         = context.NewContext()
+		ropts             = nsMetadata.Options().RetentionOptions()
```
Sure thing, thanks for the writeup too - ha.  I'll use it =]

robskillington commented on storage/bootstrap/bootstrapper/peers/source.go:
```diff
@@ -286,12 +293,23 @@ func (s *peersSource) incrementalFlush(
+
+			switch seriesCachePolicy {
+			case series.CacheAll:
+			case series.CacheAllMetadata:
```
So for CacheAllMetadata we want to retain the blocks but remove the data from memory.  For the RecentlyRead we want to remove the block completely, because we don't need it in memory at all.  

The one on the reader is completely independent to this.

richardartoul replied:
that would be dope

robskillington commented on storage/bootstrap/bootstrapper/peers/source.go:
```diff
@@ -305,22 +323,30 @@ func (s *peersSource) incrementalFlush(
+			// that we just flushed that would have to be pooled.
+			// We are explicitly trading CPU time here for lower GC pressure.
+			for _, series := range shardResultSeries {
+				bl, ok := series.Blocks.BlockAt(start)
```
This is in a large for loop, we only remove the raw data from the blocks we were successfully able to flush in this iteration of the for loop.

robskillington commented on storage/bootstrap/result/result.go:
```diff
@@ -179,6 +179,11 @@ func (sr *shardResult) BlockAt(id ts.ID, t time.Time) (block.DatabaseBlock, bool
 
+func (sr *shardResult) RemoveAll() {
+	// Faster to set to new instance of map then actually delete all entries
+	sr.blocks = make(map[ts.Hash]DatabaseSeriesBlocks)
```
Sure, can do.

robskillington commented on storage/database.go:
```diff
@@ -518,6 +518,25 @@ func (d *db) FetchBlocksMetadata(
+	shardID uint32,
+	start, end time.Time,
+	limit int64,
+	pageToken PageToken,
```
`PageToken` is just an alias for `[]byte` here.  I wanted to make page tokens opaque to the RPC layer because I need to inspect and form the response way deeper in the call stack (in the shard itself) where there is a much higher level of context about implementation details.

robskillington commented on storage/namespace.go:
```diff
@@ -511,10 +517,31 @@ func (n *dbNamespace) FetchBlocksMetadata(
+	callStart := n.nowFn()
+	shard, err := n.readableShardAt(shardID)
+	if err != nil {
+		n.metrics.fetchBlocksMetadata.ReportError(n.nowFn().Sub(callStart))
```
Not going to bother doing this, because I want to ship =]

robskillington commented on storage/namespace_readers.go:
```diff
@@ -0,0 +1,313 @@
+	blockStart time.Time,
+	position readerPosition,
+) (fs.FileSetReader, error) {
+	m.Lock()
```
Sure thing, will do.

robskillington commented on storage/namespace_readers.go:
```diff
@@ -0,0 +1,313 @@
+		m.metrics.cacheMissReusedReader.Inc(1)
+	} else {
+		reader, err = m.newReaderFn(m.bytesPool, m.fsOpts)
+		if err != nil {
```
Damn, good catch.  

Yeah it probably could.  Unfortunately we do need to unlock before calling `reader.Open(...)` because that does IO, etc.  I'll wrap this in another function that I call so I can more simply control the lock lifecycle and use defer inside that.
